drbg: Initial Commit

oddcoder opened the pull request:
This patch defines HashDrbgProps trait which holds key information about each part of the DRBG. It also implements this trait (optionally behind a feature flag) for sha2 family of hash functions from rustcrypto

Then it implements a NIST compliant hash drbg using the trait HashDrbgProps and Digest Trait from rustcrypto.

Last but not least this patch adds bunch of tests including NIST-based tests

Commit:
drbg: Initial Commit

This patch defines HashDrbgProps trait which holds key information
about each part of the DRBG. It also implements this trait (optionally
behind a feature flag) for sha2 family of hash functions from rustcrypto

Then it implements a NIST compliant hash drbg using the trait HashDrbgProps
and Digest Trait from rustcrypto.

Last but not least this patch adds bunch of tests including NIST-based
tests

Signed-off-by: Ahmed Abdelraoof <ahmed.abdelraoof@huawei.com>

oddcoder commented:
@TRJensen @trent-reed Hello again, This is follow up on #136. This patch is the real drbg implementation. As this is crypto API, I would prefer to merge this patch after approval from at least both of you. This should be fully compliant implementation of NIST DRBG. Also feel free to pull in any cryptography expert you would need to!

TRJensen commented on drbg/src/props.rs:
```diff
@@ -0,0 +1,148 @@
+            /// taken from [NIST SP 800-57 Part 1 Revision
+            /// 5](https://doi.org/10.6028/NIST.SP.800-57pt1r5) Table 3: Maximum
+            /// security strengths for hash and hash-based functions
+            const SECURITY_STRENGTH_BITS: usize = 122;
```
Typo: 122 -> 112

trent-reed replied:
+1, this should be 112 according to https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-57pt1r5.pdf

TRJensen commented on drbg/src/props.rs:
```diff
@@ -0,0 +1,148 @@
+            /// Table 2: Definitions for Hash-Based DRBG Mechanisms.
+            type SeedLenBits = $crate::digest::consts::U440;
+            type SeedLenBytes = $crate::BitsToBytes<Self::SeedLenBits>;
+            type EntropyLenBytes = $crate::digest::consts::U24;
```
Assuming this is the max entropy input length - table 2 states 2^35 bit size for all sha types - but unpractically large... We could use U32 for all definitions.

oddcoder replied:
it is true it can be any number between 0 and 2^34, but 24 bits is the number used in nist tests

TRJensen replied:
Ok, I noticed that NIST SP800-90Ar1 table2 has same max entropy size (2^35) bits ( ~ 4GB ) for all digest types. We could consider using same U24 for all digest types in our case, which allows for an entropy size of up to ~8MB bytes, which is probably is more than enough for all practical TPM use cases, but just a thought 🙂

oddcoder replied:
we cannot do that, well technically we can, but that will break NIST tests, so i picked the parameters that are tested by nist.

oddcoder commented on drbg/src/props.rs:
```diff
@@ -0,0 +1,148 @@
+            /// taken from [NIST SP 800-57 Part 1 Revision
+            /// 5](https://doi.org/10.6028/NIST.SP.800-57pt1r5) Table 3: Maximum
+            /// security strengths for hash and hash-based functions
+            const SECURITY_STRENGTH_BITS: usize = 122;
```
If I recall correctly, it was that this field has to be at least `SECURITY_STRENGTH_BITS` so there very good chance that it is bad usage of name, but I will double check

oddcoder replied:
I think this number is correct in this context, but feel free to recommend a different way to phrase it.

trent-reed replied:
I disagree, this number is supposed to represent maximum security strength FWICT from what you have here, which according to https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-57pt1r5.pdf is 112 for sha512/224.

That, or this has confused two separate people, in which case something probably needs revised.

oddcoder replied:
I checked again! good thing I didn't close the converation, but I believe you both are right. I was following ANSSi implementaton and both times I misread the constant as 122. Appologies for the trouble.

TRJensen commented on drbg/src/hashdrbg.rs:
```diff
@@ -0,0 +1,220 @@
+    }
+    /// This is the `Hash_DRBG_Generate_algorithm()` from _10.1.1.4_.
+    fn generate_inner(&mut self, additional_input: &[u8], out: &mut [u8]) {
+        let p = U896::from_u64(2).shl(Hash::SeedLenBytes::USIZE * 8);
```
We want to create the BigNum U896 having value 2^x to use for modulo operation.
We can write that:  (2<<x) = 2^(x+1) or (1<<x) = 2^x.
So more precisely we may want to do:
p = U896::from_u64(1).shl(Hash::SeedLenBytes::USIZE * 8);

TRJensen replied:
Or if allowed: U896::ONE.shl(Hash::SeedLenBytes::USIZE * 8);
(I can't deduce if the ONE constant must be copied first before shl(...), or if shl(...) will return a new instance)

oddcoder replied:
nice catch! I have no idea how I missed that and how did that pass tests!

TRJensen commented on drbg/src/hashdrbg.rs:
```diff
@@ -0,0 +1,220 @@
+        let c = slice_to_u896(&self.c);
+        let reseed_counter = U896::from_u64(self.reseed_counter);
+        // we do V = (V + H + C + reseed_counter) mod 2^seedlen in multiple steps.
+        let v = v.add_mod(&h, &p);
```
When we are doing modulo 1 << (x*8), we are basically just throwing most significant bytes away, leaving the x least significant bytes in the bignum. There is a chance we can optimize:
- if the BigNum is created to hold only (x+7)/8 limbs (u32/u64)
- the BigNum is allowed to overflow during add.
Then we just add without modulo, and return the x least significant bytes. But perhaps a future optimization as you mention above - lets leave as is for now.

oddcoder replied:
I would prefer to keep it like this to be as close as possible to specification. I will close this comment for now to help see what is left to be done.

TRJensen commented on drbg/src/hashdrbg.rs:
```diff
@@ -0,0 +1,220 @@
+    }
+    /// according to Table 2: Definitions for Hash-Based DRBG Mechanisms,
+    /// instantiate should fail if `additional_input` contains more
+    /// than `2**35` bits which is ~ 4GB.
```
Duplicate comment or missing check on additional_input size?

TRJensen commented on drbg/src/hashdrbg.rs:
```diff
@@ -0,0 +1,220 @@
+    /// instantiate should fail if `additional_input` contains more
+    /// than `2**35` bits which is ~ 4GB, we ignore that check since it is
+    /// not possible to pass that many data into a tpm chip.
+    fn next_u64(&mut self, additional_input: &[u8]) -> Result<u64, DrbgError> {
```
Duplicate comment or missing check on additional_input size?

oddcoder replied:
check is done implicitly by the the `next_u64_via_fill`

TRJensen reviewed (commented):
Looks great - some comments added 🙂👍

trent-reed commented:
Hey all! Very sorry to hit you with this - the past 2 weeks have been crazy. We lost power in large parts of my city for several days, as a result I got sick and had to recover. I have a trip planned this upcoming week, so I probably won't get to reviewing this this until *after* I return.

Sorry for the delay here, but I will prioritize it as soon as I get back from my trip (the week of the 9th).

trent-reed commented on drbg/src/hashdrbg.rs:
```diff
@@ -0,0 +1,220 @@
+pub struct HashDrbg<Hash: Digest + HashDrbgProps> {
+    /// _10.1.1.1 1. a._ A value (V) of _seedlen_ bits that is updated during
+    /// each call to the DRBG.
+    pub(crate) v: GenericArray<u8, Hash::SeedLenBytes>,
```
The only reason I can see for these being `pub(crate)` is testing. And if that's the case then the tests are in the wrong physical location (e.g. you should not need a `pub(crate)` visibility here).

A property of Rust visibility is that submodules can see the private elements of theis parent modules. So what you probably want here is this:

*  drbg/src/hashdrbg/mod.rs
*  drbg/src/hashdrbg/unit_tests.rs
*  drbg/src/hashdrbg/nist_tests.rs

And then from `mod.rs` you can declare the tests as submodules:

```
#[cfg(test)]
mod unit_tests;
#[cfg(test)]
mod nist_tests;
```

trent-reed replied:
If there's another reason I don't see then please let me know, just wanted to comment on this before I forgot.

trent-reed commented on drbg/Cargo.toml:
```diff
@@ -0,0 +1,18 @@
+[dev-dependencies]
+hex-literal = { workspace = true }
+sha1 = {workspace = true}
+sha2 = {workspace = true}
```
I think our other Cargo.toml files have a newline at the end, so maybe we should have on here just for consistency sake.

trent-reed commented on drbg/Cargo.toml:
```diff
@@ -0,0 +1,18 @@
+
+[features]
+# Implement HashDrbgProps for hash functions from sha2
+rustcrypto = ["dep:sha2"]
```
Light request for a newline between sections (e.g. a newline after the `rustcrypto = ["dep:sha2"]` and before `[dev-dependencies]`)

trent-reed commented on drbg/src/hashdrbg.rs:
```diff
@@ -0,0 +1,220 @@
+};
+use core::iter::once;
+use crypto_bigint::{prelude::ArrayEncoding, U896};
+pub use digest;
```
`pub` here basically allows this to be a part of our public API, which I'm not sure is what you want. Are you intending to re-export this crate?

trent-reed replied:
I guess since this is not a public module, it doesn't end up re-exporting. It is still a tad weird I think, I guess I don't see why this is `pub`, and would prefer leaving it non-`pub` and only occasionally use `pub` if the intent for the statement is to export from the create in the end.

So just a nit.

oddcoder replied:
Yes that is unintentional, bu I did re-export digest, it offers users of this crate access o basic utilities related to hashes.

trent-reed commented on drbg/src/lib.rs:
```diff
@@ -0,0 +1,21 @@
+#[cfg(test)]
+mod test;
+
+pub use crypto_bigint;
```
I think we should err towards not re-exporting crates unless necessary due to our public API. Is re-export of this bigint crate needed?

oddcoder replied:
So actually we do need that!

In a situation where users have acceletaors for Hash but not for hshdrbg, they can construct their own custom `Hash` types and use them for our drbg, but they will need to implement `HashDrbgProps`, and that is when they would actually need to use types from `crypto_bigint`

trent-reed commented:
I think I'm going to need to read https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-90Ar1.pdf to feel comfortable reviewing this. I'll sync back with you after I'm done. Sorry for the delay!

oddcoder commented:
> I think I'm going to need to read https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-90Ar1.pdf to feel comfortable reviewing this. I'll sync back with you after I'm done. Sorry for the delay!


Sure, no time pressure, Also I will reiterate it again, feel comfortable to pull in anyone who is cryptography expert. we definitely want to make sure that this has no unintended mistakes. 

Tomorrow, I will look at your feedback and thomas's and you will have next revision, shortly after.

Commit:
Fix comments from thomas

trent-reed commented on drbg/src/props.rs:
```diff
@@ -0,0 +1,148 @@
+use digest::{consts::U8, generic_array::ArrayLength, typenum::PartialDiv};
+
+/// Table 2: Definitions for Hash-Based DRBG Mechanisms.
```
Can we comment the section this is from for the NIST standard? e.g. "From NIST.SP.800-90Ar1: 10.1: DRBG Mechanisms Based on Hash Functions"?

oddcoder replied:
But Pretty much everything in the whole crates come from that document, I already have a top documentation stating that. I am worried that docs will be bloated, but if you prefer to have the reference placed everywhere I can do it.

trent-reed commented on drbg/src/props.rs:
```diff
@@ -0,0 +1,148 @@
+/// Do not derive this trait manually, instead use [`derive_hash_drbg_props`].
+///
+/// [`derive_hash_drbg_props`]: crate::derive_hash_drbg_props
+pub trait HashDrbgProps {
```
Slight preference for spelling the whole thing out, and using more similar terminology to the document: `HashBasedDrbgMechanism`.

trent-reed replied:
I am partially unsure if we want to allow the client to implement this themselves. If we want to lock it down we can use a sealed trait. Maybe we should start with it being sealed, and then we can unseal it later if we want to allow clients to implement this?

oddcoder replied:
I find the new name a bit confusing

According to `4 Terms and Definitions`:

```
DRBG Mechanism:
 The portion of an RBG that includes the functions necessary
to instantiate and uninstantiate the RBG, generate
pseudorandom bits, (optionally) reseed the RBG and test the
health of the the DRBG mechanism.
```

And this trait just defines bunch of constant types and constant numbers.

trent-reed commented on drbg/src/props.rs:
```diff
@@ -0,0 +1,148 @@
+    /// Seed length (seedlen) for Hash_DRBG in bits
+    type SeedLenBits: ArrayLength<u8>;
+    /// Seed length (seedlen) for Hash_DRBG converted to bytes
+    type SeedLenBytes: ArrayLength<u8>;
```
I understand why we want these, but a small part of me is a tad worried that someday there will be a technical limitation of the defined bits not being easily divided by 8 to form bytes.

I wonder if the comment should clarify that this is a floored value if there's ever a case that pops up like this (though such a case does not exist today).

oddcoder replied:
I agree with you, this can lead to errors, but Instead of tracking all of those instances, and documenting them one by one, I will modify  `BitsToBytes<T>` to represent the minimum number of bytes requires to store all needed bits. and I will create another copy for number based constants.

trent-reed commented on drbg/src/helpers.rs:
```diff
@@ -0,0 +1,41 @@
+//! Funtions here are not part of any standards, but we need them for practical reasons.
+use crypto_bigint::U896;
+use digest::generic_array::{ArrayLength, GenericArray};
```
Just a note that mid-2025 there seems to be a huge set of breaking changes to digest, part of which is moving away from `generic_array`. It won't be a hard to make change at that time, but I wanted to call it out because it'll be a good time for us to do a major version change to Rust, and edition upgrade. Additionally, we should not allow any external dependencies until this breaking change (not that this was a risk or anything, I highly doubt anything will be remotely ready by March, but I wanted to put this on your radar): https://github.com/RustCrypto/traits/issues/1717

oddcoder replied:
I wasn't aware of it, thanks, I will resolve it for now just to see the remaining issues.

trent-reed commented on drbg/src/props.rs:
```diff
@@ -0,0 +1,148 @@
+/// Do not derive this trait manually, instead use [`derive_hash_drbg_props`].
+///
+/// [`derive_hash_drbg_props`]: crate::derive_hash_drbg_props
+pub trait HashDrbgProps {
```
Why not just say that `Digest` is a supertrait requirement of this? I can't imagine we'd ever want to say there's a hash-based DRBG that cannot produce a digest. I think that would clean up the trait bounds in a lot of places so that you won't have to state `Digest + HashDrbgProps` in places, because the `Digest` will anyways be a supertrait requirement of implementing `HashDrbgProps`.

oddcoder replied:
good idea!

vsukhoml commented on drbg/src/hashdrbg.rs:
```diff
@@ -0,0 +1,220 @@
+/// ## Notes
+///  `Hash` is expected to be cryptographically secure hash function.
+///
+pub struct HashDrbg<Hash: Digest + HashDrbgProps> {
```
Would it make sense to introduce a trait for DRBG with init/generate/reseed/drop functions to allow different implementation, potentially using HW acceleration? This implementation would be one of many.
Also, for HW it would be great to make sure that there is no assumption that state can be copied.

oddcoder replied:
Actually we have two layer of traits here and you can pick which one you prefer, The first is software using Hardware `Hash` in software `HashDrbg` which just requires setting the right parameters.

The otherway is Fully hardware accelerated Drbg, in which case you don't really need this crate at all, and all you would need is the traits in the platform api from [server/src/platform/crypto/drbg/mod.rs](https://github.com/tpm-rs/tpm-rs/blob/main/server/src/platform/crypto/drbg/mod.rs)

trent-reed commented on drbg/src/props.rs:
```diff
@@ -0,0 +1,148 @@
+/// [`derive_hash_drbg_props`]: crate::derive_hash_drbg_props
+pub trait HashDrbgProps {
+    /// Seed length (seedlen) for Hash_DRBG in bits
+    type SeedLenBits: ArrayLength<u8>;
```
Part of me wants to see something a bit more structured in its definition. That way we can control this a little better and ensure that it's implemented properly, and that the meaning of the type is really clear.

Example:

```rs
pub trait BinarySize: private::Sealed {
  type Bits: ArrayLength<u8>;
  type BytesCeil: ArrayLength<u8>;
  type BytesFloor: ArrayLength<u8>;
}
```

Then, instead of using `SeedLenBits` and so-on, we'd just use `SeedLength: BinarySize`

Example Link: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=64b32f8761db6977c4ae9ee8c92e7b67

trent-reed replied:
One potential added benefit of this is we can ensure that the client doesn't implement the size calculations wrong by locking the `BinarySize` trait to only be implementable in this crate. This makes it so that we must implement all valid combinations of it for the client, reducing the chance of it being implemented improperly.

oddcoder replied:
I thought about this idea, but my problem was the main use case wouldn't work with sealed traits.

Company X has their own hash implmentation, they create `struct OurHash` and eventually  they want to use it as accelerator for the drbg. with sealed  they will not be able to fill in the parameters for their custom hash and imo that defeats the whole purpose.

trent-reed commented on drbg/src/props.rs:
```diff
@@ -0,0 +1,148 @@
+/// Do not derive this trait manually, instead use [`derive_hash_drbg_props`].
+///
+/// [`derive_hash_drbg_props`]: crate::derive_hash_drbg_props
+pub trait HashDrbgProps {
```
I am partially unsure if we want to allow the client to implement this themselves. If we want to lock it down we can use a sealed trait. Maybe we should start with it being sealed, and then we can unseal it later if we want to allow clients to implement this?

trent-reed commented on drbg/src/props.rs:
```diff
@@ -0,0 +1,148 @@
+    /// the numbers are taken from NIST test Vectors
+    type NonceLenBytes: ArrayLength<u8>;
+    /// highest_supported_security_strength
+    const SECURITY_STRENGTH_BITS: usize;
```
We should probably be consistent an use the same mechanism here as we do for the other bits/bytes calculations. `ArrayLength` implements `Unsigned`, so we can anyways `.to_usize()` or whatever if they need a usize.

trent-reed replied:
I would do what makes the most sense. If we don't find using `usize` directly to be confusing, then we can just continue using it. We can stick with that for now, as this is only a nit-pick anyways.

trent-reed commented on drbg/src/props.rs:
````diff
@@ -0,0 +1,148 @@
+///
+/// // now you can use HashDrbg<MyCustomSha256>
+/// ```
+#[macro_export]
````
If we want to control this with a sealed trait for the time being, we won't need to export this. If we want to allow external clients to implement this though, then we'd still need it. Just mentioning here in case you opt for the more restrictive option of sealing.

trent-reed commented on drbg/src/props.rs:
```diff
@@ -0,0 +1,148 @@
+    /// Seed length (seedlen) for Hash_DRBG converted to bytes
+    type SeedLenBytes: ArrayLength<u8>;
+    /// the numbers are taken from NIST test Vectors
+    type EntropyLenBytes: ArrayLength<u8>;
```
I don't know that I like these `EntropyLen` and `NonceLen` associated types.

The actual definition of the required entropy according to the table we are referencing is `security_strength`, which is a measurement of how the DRBG is *intended to be used*, not some inherent property of the DRBG implementation.

For example, if I have a SHA-512 DRBG (≥ 256 bits of security strength, according to https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-57pt1r5.pdf), and I want to use that for something that requires only 192 bits of security - I totally can, and that basically changes the required nonce length and entropy length for my specific use-case.

What putting these on the trait does is suggest something inherent about the DRBG implementation, but that sounds incorrect from my reading of section 10.

oddcoder replied:
we can revisit before merging after addressing all other issues

trent-reed commented on drbg/src/props.rs:
```diff
@@ -0,0 +1,148 @@
+    /// the numbers are taken from NIST test Vectors
+    type NonceLenBytes: ArrayLength<u8>;
+    /// highest_supported_security_strength
+    const SECURITY_STRENGTH_BITS: usize;
```
I can see the desire for something like this, unlike the desire for nonce or entropy requirements, which depend more on the use-case.

However, I think we should be very clear that it's a *max*. So we should say something like `MaxSecurityStrength` or something.

trent-reed replied:
Actually, the spec uses the term "highest", not "max" (see: `highest_supported_security_strength`). Probably we should follow that wording.

trent-reed commented on drbg/src/props.rs:
```diff
@@ -0,0 +1,148 @@
+    const SECURITY_STRENGTH_BITS: usize;
+    const SECURITY_STRENGTH_BYTES: usize = Self::SECURITY_STRENGTH_BITS / 8;
+    /// 8.6.7.a Nonce: A value with at least (security_strength/2) bits of entropy
+    const NONCE_BYTES_MIN: usize = Self::SECURITY_STRENGTH_BYTES / 2;
```
Much like my other comments above, I just don't know that this makes sense. The nonce bytes required depend on the use case in reality.

oddcoder replied:
it is true, some constants have range, and I chose to fix them to the parameters specifid by nist test vectors, we can revisit this one later before merging to see how much complexity will it add.

trent-reed commented on drbg/src/props.rs:
```diff
@@ -0,0 +1,148 @@
+/// Do not derive this trait manually, instead use [`derive_hash_drbg_props`].
+///
+/// [`derive_hash_drbg_props`]: crate::derive_hash_drbg_props
+pub trait HashDrbgProps {
```
Kinda wondering why we don't have `OutputBlockLength` here. It seems like an important property that varies an amount that is defined by the spec.

oddcoder replied:
It is true, it is defined by the spec, but that is a property of the hash function that we do not need to use manually. It is the size of the output which is always known at compile time.

trent-reed commented on drbg/src/props.rs:
```diff
@@ -0,0 +1,148 @@
+
+/// Table 2: Definitions for Hash-Based DRBG Mechanisms:
+/// `max_number_of_bits_per_request`
+pub const MAX_BITS_PER_REQUEST: usize = 1 << 19;
```
The spec says "2^19", this is "1^19", unless I'm reading it wrong.

trent-reed replied:
I originally read this wrong, I suspect that some of these numbers are wrong here though, because some of the numbers use 2 but this number uses 1, and they all use shifts. They are all defined in the spec as `2^N`, which the equivalent of in shift notation would be either `1<<N` or `2<<(N-1)`.

Anyways, a comment something along the lines of the notation that the spec uses would probably be nice. And we should canonicalize all of these to do the same-ish thing, which means some of the values below are probably incorrect.

trent-reed commented on drbg/src/props.rs:
```diff
@@ -0,0 +1,148 @@
+
+/// Table 2: Definitions for Hash-Based DRBG Mechanisms:
+/// `max_number_of_bits_per_request`
+pub const MAX_BITS_PER_REQUEST: usize = 1 << 19;
```
I do wonder if these should be properties on the trait for consistency, despite them being identical across all implementations.

oddcoder replied:
I don't have preference, after all they will be derived automatically by the macro.

trent-reed commented on drbg/src/hashdrbg.rs:
```diff
@@ -0,0 +1,220 @@
+use crate::{
+    helpers::{slice_to_u896, truncate_from_start},
+    props::{
+        MAX_BYTES_ADDITIONAL_INPUT, MAX_BYTES_PERSONALIZATION_STRING, MAX_BYTES_PER_REQUEST,
```
Not crazy about these constant names. They are specific to hash-based DRBG, but they don't suggest as much with their name. CTR DRBG has different values for these in some cases.

oddcoder replied:
At least at this state, this crate is just hashdrbbg. At least for my side, I have no intnetion on adding any other implementation. maybe we should rename the crate to hashdrbg instead of drbg ?

trent-reed commented on drbg/src/hashdrbg.rs:
```diff
@@ -0,0 +1,220 @@
+    fn hash_df(data: &[&[u8]]) -> GenericArray<u8, Hash::SeedLenBytes> {
+        let mut tmp = GenericArray::default();
+        let mut counter: u8 = 1;
+        let len = Hash::SeedLenBytes::USIZE / Hash::OutputSize::USIZE;
```
Technically this is defined as a measurement of the bits in the spec, and we take the ceil of the result, but this calculation here is over the bytes, and takes the floor of the result.

Since all of the data is divisible by 8 evenly, this is okay, but it's technically depending on that fact. Maybe we should add a `static_assert` of some kind here with a comment describing why that's okay?

trent-reed replied:
You do account for this later, I just noticed; I think maybe we should lean into the 8 division though, since no current implementation doesn't divide evenly, and just make a `static_assert` that ensures that's the case. It's hard to see that something does what is expected if it isn't really used anywhere.

trent-reed commented on drbg/src/hashdrbg.rs:
```diff
@@ -0,0 +1,220 @@
+        let mut tmp = GenericArray::default();
+        let mut counter: u8 = 1;
+        let len = Hash::SeedLenBytes::USIZE / Hash::OutputSize::USIZE;
+        let no_of_bits_to_return: &[u8] = &(Hash::SeedLenBytes::U32 * 8).to_be_bytes();
```
I wonder if we should use `SeedLenBits` instead for clarity?

trent-reed commented on drbg/src/hashdrbg.rs:
```diff
@@ -0,0 +1,220 @@
+        let mut tmp = GenericArray::default();
+        let mut counter: u8 = 1;
+        let len = Hash::SeedLenBytes::USIZE / Hash::OutputSize::USIZE;
+        let no_of_bits_to_return: &[u8] = &(Hash::SeedLenBytes::U32 * 8).to_be_bytes();
```
technically could be `const`, I think: `const NO_OF_BITS_TO_RETURN: [u8; 4] = Hash::SeedLenBits::U32.to_be_bytes()`

oddcoder replied:
It required type annotation for chaining, I prefer not to have to worry about those.

trent-reed commented on drbg/src/hashdrbg.rs:
```diff
@@ -0,0 +1,220 @@
+        let mut counter: u8 = 1;
+        let len = Hash::SeedLenBytes::USIZE / Hash::OutputSize::USIZE;
+        let no_of_bits_to_return: &[u8] = &(Hash::SeedLenBytes::U32 * 8).to_be_bytes();
+        for i in 0..len {
```
Optional: To make it more in-line with the wording of the spec, we could make it an inclusive loop: `for _ in 1..=len`.

trent-reed commented on drbg/src/hashdrbg.rs:
```diff
@@ -0,0 +1,220 @@
+            let out = Self::hash(hashables);
+            tmp[i * Hash::OutputSize::USIZE..(i + 1) * Hash::OutputSize::USIZE]
+                .copy_from_slice(&out);
+            counter += 1;
```
Counter is basically identical to `i` if we set the range to `1..=len`, so maybe we should just do that and then call `i` `counter` instead.

trent-reed commented on drbg/src/hashdrbg.rs:
```diff
@@ -0,0 +1,220 @@
+                .chain(once(&no_of_bits_to_return))
+                .chain(data.iter());
+            let out = Self::hash(hashables);
+            tmp[i * Hash::OutputSize::USIZE..(i + 1) * Hash::OutputSize::USIZE]
```
I wonder if it'd be cleaner/easier to use `chunks_mut` or `chunks_exact_mut` instead of doing this chunking math inline?

trent-reed replied:
I still think this is easier to read. See:

```rs
fn hash_df(data: &[&[u8]]) -> GenericArray<u8, Hash::SeedLenBytes> {
        let no_of_bits_to_return: &[u8] = &Hash::SeedLenBits::U32.to_be_bytes();
        let mut result = GenericArray::default();
        for (idx, chunk) in result.chunks_mut(Hash::OutputSize::USIZE).enumerate() {
          let header = [
            &[(idx as u8) + 1],   // Counter byte (starts from 1)
            no_of_bits_to_return,
          ];
          let hashables = header.iter().chain(data.iter());
          let out = Self::hash(hashables);
          chunk.copy_from_slice(&out[..chunk.len()]);
        }
        result
}
```

trent-reed replied:
In general, I'm very anti-"calling the [] operator" in Rust.

trent-reed replied:
This is getting confusing to review. I'm going to close these comments and open them again at the new code location.

trent-reed commented on drbg/src/hashdrbg.rs:
```diff
@@ -0,0 +1,220 @@
+    /// This function omits the number of bits to return, it always returns an array of
+    /// the length [HashDrbgProps::SeedLenBytes]
+    fn hash_df(data: &[&[u8]]) -> GenericArray<u8, Hash::SeedLenBytes> {
+        let mut tmp = GenericArray::default();
```
Technically this is memzero-ing every call. I think that's desired, but I want to be clear about it in case it's not for these low-power devices. If not, we may have to dip into unsafe.

oddcoder replied:
I was not thinking about that honeslty when I was writing code.

trent-reed commented on drbg/src/hashdrbg.rs:
```diff
@@ -0,0 +1,220 @@
+
+impl<Hash: Digest + HashDrbgProps> HashDrbg<Hash> {
+    /// Generic hash function defined by `Hash` type
+    fn hash(data: impl Iterator<Item = impl AsRef<[u8]>>) -> Output<Hash> {
```
Instead of taking an iterator, just take a slice and iterate over that IMO. No need to gen a new function per call, and it'll probably clean up your callsites pretty nicely I'd think.

trent-reed commented on drbg/src/hashdrbg.rs:
```diff
@@ -0,0 +1,220 @@
+        let no_of_bits_to_return: &[u8] = &(Hash::SeedLenBytes::U32 * 8).to_be_bytes();
+        for i in 0..len {
+            let counter_byte: &[u8] = &[counter];
+            let hashables = once(&counter_byte)
```
Optional: Might be a tad easier to say that `hash` is over `impl Iterator<Item=u8>`, then you can flatten your iterator chains a bit: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=557710a5cbfd0e2d8ab1e991cf5ecda4

trent-reed commented on drbg/src/hashdrbg.rs:
```diff
@@ -0,0 +1,220 @@
+    ///
+    /// This function omits the number of bits to return, it always returns an array of
+    /// the length [HashDrbgProps::SeedLenBytes]
+    fn hash_df(data: &[&[u8]]) -> GenericArray<u8, Hash::SeedLenBytes> {
```
For ease of equivalence, I wonder if this should also be `impl Iterator<Item=u8>` instead of having this be a slice of slices (which is kind of the intent for how it's used leaking into the function signature).

trent-reed replied:
I'm on the fence here, it is nice to see `Self::hash_df(&[A, B, C, ...])`; so maybe what you have is fine.

oddcoder replied:
In this case I will resolve this for now, feel free to re-opon

trent-reed commented on drbg/src/hashdrbg.rs:
```diff
@@ -0,0 +1,220 @@
+
+impl<Hash: Digest + HashDrbgProps> HashDrbg<Hash> {
+    /// Generic hash function defined by `Hash` type
+    fn hash(data: impl Iterator<Item = impl AsRef<[u8]>>) -> Output<Hash> {
```
Optional: Might be a tad easier to say that `hash` is over `impl Iterator<Item=u8>`, then you can flatten your iterator chains a bit: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=557710a5cbfd0e2d8ab1e991cf5ecda4

oddcoder replied:
the problem is with `Digest::update()`. It roughly, expects &[u8] so if we do `Iterator` we need to collect, and I think that is undesirable.

trent-reed commented on drbg/src/hashdrbg.rs:
```diff
@@ -0,0 +1,220 @@
+                .copy_from_slice(&out);
+            counter += 1;
+        }
+        let remainder = Hash::SeedLenBytes::USIZE % Hash::OutputSize::USIZE;
```
Ah, I didn't realize you handled this - Maybe instead we should start with a `static_assert` of some kind that they divide equally for now, and just punt this off to if we ever need to implement this in the future. I'd rather do that, personally.

oddcoder replied:
this is the relation between seed length and output length, remainder, does not have to be zero. I am not sure I understand your suggestion.

trent-reed commented on drbg/src/hashdrbg.rs:
```diff
@@ -0,0 +1,220 @@
+        self.reseed_counter = 1;
+    }
+    /// This is the `Hash_DRBG_Generate_algorithm()` from _10.1.1.4_.
+    fn generate_inner(&mut self, additional_input: &[u8], out: &mut [u8]) {
```
It worries me that this cannot fail, when the specification states that it can and should fail if the `reseed_counter > reseed_interval`; am I missing something here?

oddcoder replied:
good, point which means it needs some comments, I try to separate concerns.

This method, among others, have assumptions that must be fulfilled by the caller.

this method assumes `out <MAX_BYTES_PER_REQUEST` and `additional_input < MAX_BYTES_ADDITIONAL_INPUT` and that `reseed_counter < MAX_REQUESTS_BETWEEN_RESEEDS`. The single caller of this function `fill_bytes` checks those.

I will document all assumed invariants.

trent-reed commented:
General comment: I wonder why we are implementing this. I suspect that nobody is going to want to use this, and instead would rather use some other implementation (e.g. OpenSSL-provided one, or otherwise). I strongly suspect we don't want to implement this, and instead we can bind the OSSL one in as a placeholder that folks can replace.

trent-reed commented on drbg/src/hashdrbg.rs:
```diff
@@ -0,0 +1,220 @@
+    fn generate_inner(&mut self, additional_input: &[u8], out: &mut [u8]) {
+        let p = U896::from_u64(2).shl(Hash::SeedLenBytes::USIZE * 8);
+        let mut v = slice_to_u896(&self.v);
+        if !additional_input.is_empty() {
```
As a point of comparison, here is the "conversion to V" that OpenSSL implements: https://github.com/freebsd/freebsd-src/blob/main/crypto/openssl/providers/implementations/rands/drbg_hash.c#L175-L187

trent-reed commented:
> General comment: I wonder why we are implementing this. I suspect that nobody is going to want to use this, and instead would rather use some other implementation (e.g. OpenSSL-provided one, or otherwise). I strongly suspect we don't want to implement this, and instead we can bind the OSSL one in as a placeholder that folks can replace.

Before I continue reviewing this, I think I'd like an answer to this question. I suspect we should not be implementing this, and instead should - if we do anything - bind in OpenSSL and provide that as a "here's an example implementation". Maybe I'm missing something, but that's really what I would have done here.

Commit:
Fixing spelling mistake in drbg constants

Apparntly I missed it twice, rereading it as 122 every time.

Signed-off-by: Ahmed Abdelraoof <ahmed.abdelraoof@huawei.com>

Commit:
fix cargo.toml issues

Signed-off-by: Ahmed Abdelraoof <ahmed.abdelraoof@huawei.com>

Commit:
removing unnecessary re-export

Signed-off-by: Ahmed Abdelraoof <ahmed.abdelraoof@huawei.com>

Commit:
fix spelling mistake

Signed-off-by: Ahmed Abdelraoof <ahmed.abdelraoof@huawei.com>

Commit:
change bit to bytes calculation to be more accurate

Signed-off-by: Ahmed Abdelraoof <ahmed.abdelraoof@huawei.com>

Commit:
Make digest subtrait of HashDrbgProps

Signed-off-by: Ahmed Abdelraoof <ahmed.abdelraoof@huawei.com>

Commit:
fixup! Make digest subtrait of HashDrbgProps

Signed-off-by: Ahmed Abdelraoof <ahmed.abdelraoof@huawei.com>

oddcoder commented:
> Before I continue reviewing this, I think I'd like an answer to this question. I suspect we should not be implementing this, and instead should - if we do anything - bind in OpenSSL and provide that as a "here's an example implementation". Maybe I'm missing something, but that's really what I would have done here.

@trent-reed we had this discussion before i started implementing this patch and @chrisfenner's guidance was to prefer having rust based one instead of linking against openssl.

Also I was slightly in favor of exclusively using rustcrypto /rand and expanding on them when needed (as in the case of the drbg)

Commit:
Rename variables to be more compliant with nist

SECURITY_STRENGTH_BITS 		         and SECURITY_STRENGTH_BYTES to
HIGHEST_SUPPORTED_SECURITY_STRENGTH_BITS and HIGHEST_SUPPORTED_SECURITY_STRENGTH_BYTES

Commit:
Small nits

Signed-off-by: Ahmed Abdelraoof <ahmed.abdelraoof@huawei.com>

Commit:
remove pub(crate)

Signed-off-by: Ahmed Abdelraoof <ahmed.abdelraoof@huawei.com>

Commit:
Documnt the invariants of critical functions here

Signed-off-by: Ahmed Abdelraoof <ahmed.abdelraoof@huawei.com>

oddcoder commented:
@vsukhoml @kstruczy @TRJensen @trent-reed I resolved most issues, feel free to close conversations that are resolved or emphasize problems that you think are still there.

Note that next two weeks, I will be changing cities and dealing with relocation-related real world action, so I will be a bit slow to respond.

also happy new year to everyone!

oddcoder commented:
Ping @trent-reed @vsukhoml @TRJensen @kstruczy ?

trent-reed commented:
Sorry for the delay, I've been in the weeds on another project still. I'll schedule some time to look at this again on Monday.

trent-reed reviewed (commented):
Sorry for the delay, here's the next pass!

oddcoder commented:
sorry for the long delay! I am literally scrapping or free 10 minutes here and there. The upside is I forgot my code, so I will be looking at it as if it was written by a total stranger, I will look into your comments this weekend, I will close this CR and open new one to have differentiate new from old comments.

oddcoder commented:
Hi, I think I have no capacity for next few month to work on tpm-rs, onboarding/ops work is leaving zero mental capacity for side projects I hope it gets better with time, But I prefer not to merge it knowing I cannot maintain for a while, I you are ok with that let me know and I will hit merge

trent-reed commented:
Hmm, we'll make a decision on this in the next meeting - thanks for letting us know!
